### ISSUE TITLE
Not generating C++/CLI interface class from C++ interface

### ISSUE THREAD

**Issue by inkadnb** [mono/CppSharp#721]
##### Brief Description
Given a C++ interface like the one below

```
class IShape {
public:
  virtual ~IShape();
  virtual void move_x(distance x) = 0;
  virtual void move_y(distance y) = 0;
  virtual void rotate(angle rotation) = 0;
  //...
};
```

CppSharp will still generate a class, not an interface class.

OS: Windows / OS X / Linux (include version and/or distro)
Windows 10

##### Used headers
CppSharp.h

##### Used settings
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CppSharp;
using CppSharp.AST;
using CppSharp.Parser;
using GenerateBindings;
using CppSharp.Generators;
using System.IO;
using CppSharp.Passes;
namespace GenerateBindings
{
    class Library : ILibrary
    {
        public void Postprocess(Driver driver, ASTContext ctx)
        {
        }
        public void Preprocess(Driver driver, ASTContext ctx)
        {
            ctx.IgnoreHeadersWithName("wglext.h");
            ctx.IgnoreHeadersWithName("qikmath.h");
            ctx.RenameNamespace("qgameengine", "QuadrionEngine");
        }
        public void Setup(Driver driver)
        {
            string path = Directory.GetCurrentDirectory();
            Console.WriteLine(path);
            var options = driver.Options;
            //driver.ParserOptions.AddDefines("QGAMEENGINE_EXPORTS");
            //driver.ParserOptions.AddDefines("QENGINE_EXPORTS");
            //driver.ParserOptions.AddSourceFiles(@"C:\Users\jonth\Documents\Quad2012\projects\qengine\src");
            //driver.ParserOptions.AddSourceFiles(@"C:\Users\jonth\Documents\Quad2012\projects\qgameengine\src");
            driver.ParserOptions.AddIncludeDirs(@"C:\Users\jonth\Documents\Quad2012\projects\qengine\include");
            driver.ParserOptions.AddIncludeDirs(@"C:\Users\jonth\Documents\Quad2012\projects\qgameengine\include");
            driver.ParserOptions.AddIncludeDirs(@"C:\Users\jonth\Documents\Quad2012\ext\GL");
            driver.ParserOptions.AddSystemIncludeDirs(@"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include");
            driver.ParserOptions.AddSystemIncludeDirs(@"c:\Program Files (x86)\Windows Kits\8.1\Include\um");
            driver.ParserOptions.AddSystemIncludeDirs(@"c:\Program Files (x86)\Windows Kits\8.1\Include\shared");
            options.GeneratorKind = GeneratorKind.CLI;
            options.LibraryName = "QuadrionEngine";
            options.Headers.Add(@"C:\Users\jonth\Documents\Quad2012\projects\qgameengine\include\qcomponent.h");
            options.OutputDir = "ManagedBindings";
            options.CompileCode = true;
        }
        public void SetupPasses(Driver driver)
        {
        }
    }
}
```

**Comment by ddobrev** [comment]
Given the name of the original C++ class/interface, I assume you do not want a class at all - just an interface. This cannot happen automatically because, as you correctly noted, C++ has no notion of interfaces. What we can do is add a new feature to allow users to manually adjust certain classes to be interfaces instead. In the mean time, if the C# generator works for you, you could switch to that because if your interfaces happen to be used in multiple inheritances, the C# generator is going to detect them and add interfaces to complement your classes. The only minor problem you're going to have is that your interfaces will start with "II" instead of just "I", for example - IIShape.

**Comment by inkadnb** [comment]
> Given the name of the original C++ class/interface, I assume you do not want a class at all - just an interface.

Correct.

> In the mean time, if the C# generator works for you, you could switch to that because if your interfaces happen to be used in multiple inheritances, the C# generator is going to detect them and add interfaces to complement your classes.

I haven't tried to use the C# generator yet. Is it still able to call back into the native code? I ask because my plan is to instance the CLR within C++, load in a .NET assembly that references the C++/CLI assembly and instance a class that-way but due to the nature of these scripts I need to be able to marshal in both directions.

> The only minor problem you're going to have is that your interfaces will start with "II" instead of just "I", for example - IIShape.
That's not a major issue to me, definitely something that can be fixed by just changing up the naming conventions in the C++ code, or adding a step in the generator.

**Comment by tritao** [comment]
Yeah this should be relatively easy to implement. There are a couple ways to do it.

1. You could add a new pass, analogous to `CheckStaticClass` (see https://github.com/mono/CppSharp/blob/master/src/Generator/Passes/CheckStaticClass.cs) that will check if the class is an interface with some heuristics, for example if it only contains pure virtual methods) and switch the class to be an interface.

2. Extend the `CheckMacrosPass` pass (see https://github.com/mono/CppSharp/blob/master/src/Generator/Passes/CheckMacrosPass.cs) with a new `CS_INTERFACE` that also changes to the class to be an interface.

3. In `Preprocess` change the class manually to be an interface (using the `Type` property in class, which would take a `ClassType.Interface` value in your case.

It should be quite easy to do it, if you want to give it a try.

**Comment by tritao** [comment]
By the way, it should be possible to replace your manual VS / Windows SDK header setup with the `ParserOptions.SetupMSVC(VisualStudioVersion)` method. If it doesn't work, let us know.
